docs/design/CONFIG_API.md

```diff
@@ -0,0 +1,330 @@
+Configuration API Design
+===================
+
+## Motivation
+
+[Issue 51][issue_51] specifies the need to design an API for configuring Envoy Gateway at runtime. This configuration is
+referred to as the "Bootstrap Config" in the Envoy Gateway high-level [design doc][design_doc]. This document replaces
+the term "bootstrap" with "runtime" as some users may confuse bootstrap with Envoy's [bootstrap][envoy_boot]
+configuration.
+
+## Goals
+
+* Define an API for configuring Envoy Gateway, independent of the runtime platform.
+* Specify tooling for managing the API, e.g. generate protos, CRDs, controller RBAC, etc.
+
+## Non-Goals
+
+* Implementation of the API.
+* Define the `status` subresource of the configuration API.
+* Define a complete configuration API for managing Envoy Gateway at runtime. The configuration API will evolve over
+  time.
+* Define a complete configuration API for managing control and data planes. The configuration API will evolve over time.
+
+## Proposal
+
+Utilize the [Kubebuilder][kubebuilder] framework to build and manage Envoy Gateway APIs. Kubebuilder facilitates the
+following developer workflow for building APIs:
+
+- Create one or more resource APIs as CRDs and then add fields to the resources.
+- Implement reconcile loops in controllers and watch additional resources.
+- Test by running against a cluster (self-installs CRDs and starts controllers automatically).
+- Update bootstrapped integration tests to test new fields and business logic.
+- Build and publish a container from a Dockerfile.
+
+After installing the latest Kubebuilder release, create the `RuntimeConfig` API:
+```shell
+kubebuilder create api --group gateway.envoyproxy.io --version v1alpha1 --kind RuntimeConfig
+```
+
+The `v1alpha1` version and `gateway.envoyproxy.io` API group get generated:
+```go
+// gateway/api/v1alpha1/doc.go
+
+// Package v1alpha1 contains API Schema definitions for the gateway.envoyproxy.io API group.
+//
+// +groupName=gateway.envoyproxy.io
+package v1alpha1
+```
+
+The initial `RuntimeConfig` API being proposed:
+```go
+// gateway/api/v1alpha1/runtimeconfig_types.go
+
+// RuntimeConfigSpec defines the desired state of RuntimeConfig.
+type RuntimeConfigSpec struct {
+	// Providers provide runtime configuration. If unspecified, the Kubernetes
+	// provider will be used with default parameters.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=4
+	// +kubebuilder:default={kubernetes: {foo: bar}}
+	Providers []ProviderConfig `json:"providers,omitempty"`
+
+	// ControlPlane defines control plane configuration parameters. If unspecified,
+	// the control plane will use default configuration parameters.
+	//
+	// +optional
+	ControlPlane *ControlPlaneConfig `json:"controlPlane,omitempty"`
+
+	// DataPlane defines data plane configuration parameters. If unspecified,
+	// the data plane will use default configuration parameters.
+	//
+	// +optional
+	DataPlane *DataPlaneConfig `json:"dataPlane,omitempty"`
```

Comment:
im guessing DataPlane will contain platform agnostic fields, so if the user wanted to specify a service type for the envoy k8s service, that would be expressed using the top level Provider field ?

Reply:
@arkodg ControlPlane and DataPlane APIs expose provider-agnostic configuration. For example, `spec.dataPlane.adminAddress` and `spec.dataPlane.adminPort` could be used to configure Envoy's admin endpoint. I'm open to removing data plane configuration from the design spec until it's needed.